// @theia/core/lib/browser/preferences/preference-service.js
"use strict";
/********************************************************************************
 * Copyright (C) 2018 Ericsson and others.
 *
 * This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License v. 2.0 which is available at
 * http://www.eclipse.org/legal/epl-2.0.
 *
 * This Source Code may also be made available under the following Secondary
 * Licenses when the conditions for such availability set forth in the Eclipse
 * Public License v. 2.0 are satisfied: GNU General Public License, version 2
 * with the GNU Classpath Exception which is available at
 * https://www.gnu.org/software/classpath/license.html.
 *
 * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
 ********************************************************************************/
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __values = (this && this.__values) || function (o) {
    var m = typeof Symbol === "function" && o[Symbol.iterator], i = 0;
    if (m) return m.call(o);
    return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
};
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
Object.defineProperty(exports, "__esModule", { value: true });
// tslint:disable:no-any
var inversify_1 = require("inversify");
var coreutils_1 = require("@phosphor/coreutils");
var common_1 = require("../../common");
var promise_util_1 = require("../../common/promise-util");
var preference_contribution_1 = require("./preference-contribution");
var uri_1 = require("../../common/uri");
var preference_scope_1 = require("./preference-scope");
exports.PreferenceScope = preference_scope_1.PreferenceScope;
var PreferenceChangeImpl = /** @class */ (function () {
    function PreferenceChangeImpl(change, providers) {
        this.change = change;
        this.providers = providers;
    }
    Object.defineProperty(PreferenceChangeImpl.prototype, "preferenceName", {
        get: function () {
            return this.change.preferenceName;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PreferenceChangeImpl.prototype, "newValue", {
        get: function () {
            return this.change.newValue;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PreferenceChangeImpl.prototype, "oldValue", {
        get: function () {
            return this.change.oldValue;
        },
        enumerable: true,
        configurable: true
    });
    PreferenceChangeImpl.prototype.affects = function (resourceUri) {
        var e_1, _a;
        if (this.change.domain && resourceUri &&
            this.change.domain.length !== 0 &&
            this.change.domain.map(function (uriStr) { return new uri_1.default(uriStr); })
                .every(function (folderUri) { return folderUri.path.relativity(new uri_1.default(resourceUri).path) < 0; })) {
            return false;
        }
        var _loop_1 = function (scope, provider) {
            if (!resourceUri && scope === preference_scope_1.PreferenceScope.Folder) {
                return "continue";
            }
            var providerInfo = provider.canProvide(this_1.preferenceName, resourceUri);
            var priority = providerInfo.priority;
            if (priority >= 0 && scope > this_1.change.scope) {
                return { value: false };
            }
            if (scope === this_1.change.scope && this_1.change.domain.some(function (d) { return providerInfo.provider.getDomain().findIndex(function (pd) { return pd === d; }) < 0; })) {
                return { value: false };
            }
        };
        var this_1 = this;
        try {
            for (var _b = __values(this.providers.entries()), _c = _b.next(); !_c.done; _c = _b.next()) {
                var _d = __read(_c.value, 2), scope = _d[0], provider = _d[1];
                var state_1 = _loop_1(scope, provider);
                if (typeof state_1 === "object")
                    return state_1.value;
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
            }
            finally { if (e_1) throw e_1.error; }
        }
        return true;
    };
    return PreferenceChangeImpl;
}());
exports.PreferenceChangeImpl = PreferenceChangeImpl;
exports.PreferenceService = Symbol('PreferenceService');
/**
 * We cannot load providers directly in the case if they depend on `PreferenceService` somehow.
 * It allows to load them lazilly after DI is configured.
 */
exports.PreferenceProviderProvider = Symbol('PreferenceProviderProvider');
var PreferenceServiceImpl = /** @class */ (function () {
    function PreferenceServiceImpl() {
        this.onPreferenceChangedEmitter = new common_1.Emitter();
        this.onPreferenceChanged = this.onPreferenceChangedEmitter.event;
        this.onPreferencesChangedEmitter = new common_1.Emitter();
        this.onPreferencesChanged = this.onPreferencesChangedEmitter.event;
        this.toDispose = new common_1.DisposableCollection(this.onPreferenceChangedEmitter, this.onPreferencesChangedEmitter);
        this.providers = [];
        this.providersMap = new Map();
        /**
         * @deprecated Use getPreferences() instead
         */
        this.preferences = {};
        this._ready = new promise_util_1.Deferred();
    }
    PreferenceServiceImpl.prototype.init = function () {
        var _this = this;
        this.toDispose.push(common_1.Disposable.create(function () { return _this._ready.reject(); }));
        this.doSetProvider(preference_scope_1.PreferenceScope.Default, this.schema);
    };
    PreferenceServiceImpl.prototype.dispose = function () {
        this.toDispose.dispose();
    };
    Object.defineProperty(PreferenceServiceImpl.prototype, "ready", {
        get: function () {
            return this._ready.promise;
        },
        enumerable: true,
        configurable: true
    });
    PreferenceServiceImpl.prototype.initialize = function () {
        this.initializeProviders();
    };
    PreferenceServiceImpl.prototype.initializeProviders = function () {
        var _this = this;
        var e_2, _a;
        try {
            this.createProviders();
            if (this.toDispose.disposed) {
                return;
            }
            try {
                for (var _b = __values(this.providersMap.values()), _c = _b.next(); !_c.done; _c = _b.next()) {
                    var provider = _c.value;
                    this.toDispose.push(provider.onDidPreferencesChanged(function (changes) {
                        return _this.reconcilePreferences(changes);
                    }));
                }
            }
            catch (e_2_1) { e_2 = { error: e_2_1 }; }
            finally {
                try {
                    if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
                }
                finally { if (e_2) throw e_2.error; }
            }
            Promise.all(this.providers.map(function (p) { return p.ready; })).then(function () { return _this._ready.resolve(); });
        }
        catch (e) {
            this._ready.reject(e);
        }
    };
    PreferenceServiceImpl.prototype.createProviders = function () {
        var _this = this;
        var providers = [];
        preference_scope_1.PreferenceScope.getScopes().forEach(function (scope) {
            var p = _this.doCreateProvider(scope);
            if (p) {
                providers.push(p);
            }
        });
        return providers;
    };
    PreferenceServiceImpl.prototype.reconcilePreferences = function (changes) {
        var _this = this;
        var e_3, _a, e_4, _b, e_5, _c;
        var changesToEmit = {};
        var acceptChange = function (change) {
            return _this.getAffectedPreferenceNames(change, function (preferenceName) {
                return changesToEmit[preferenceName] = new PreferenceChangeImpl(__assign({}, change, { preferenceName: preferenceName }), _this.providersMap);
            });
        };
        if (changes) {
            try {
                for (var _d = __values(Object.keys(changes)), _e = _d.next(); !_e.done; _e = _d.next()) {
                    var preferenceName = _e.value;
                    var change = changes[preferenceName];
                    if (!change.newValue) {
                        var overriden = this.overridenPreferenceName(change.preferenceName);
                        if (overriden) {
                            change = __assign({}, change, { newValue: this.doGet(overriden.preferenceName) });
                        }
                    }
                    if (this.schema.isValidInScope(preferenceName, preference_scope_1.PreferenceScope.Folder)) {
                        acceptChange(change);
                        continue;
                    }
                    try {
                        for (var _f = __values(preference_scope_1.PreferenceScope.getReversedScopes()), _g = _f.next(); !_g.done; _g = _f.next()) {
                            var scope = _g.value;
                            if (this.schema.isValidInScope(preferenceName, scope)) {
                                var provider = this.getProvider(scope);
                                if (provider) {
                                    var value = provider.get(preferenceName);
                                    if (scope > change.scope && value !== undefined && value !== null) {
                                        // preference defined in a more specific scope
                                        break;
                                    }
                                    else if (scope === change.scope) {
                                        acceptChange(change);
                                    }
                                }
                            }
                        }
                    }
                    catch (e_4_1) { e_4 = { error: e_4_1 }; }
                    finally {
                        try {
                            if (_g && !_g.done && (_b = _f.return)) _b.call(_f);
                        }
                        finally { if (e_4) throw e_4.error; }
                    }
                }
            }
            catch (e_3_1) { e_3 = { error: e_3_1 }; }
            finally {
                try {
                    if (_e && !_e.done && (_a = _d.return)) _a.call(_d);
                }
                finally { if (e_3) throw e_3.error; }
            }
        }
        else { // go through providers for the Default, User, and Workspace Scopes to find delta
            var newPrefs = this.getPreferences();
            var oldPrefs = this.preferences;
            try {
                for (var _h = __values(Object.keys(newPrefs)), _j = _h.next(); !_j.done; _j = _h.next()) {
                    var preferenceName = _j.value;
                    var newValue = newPrefs[preferenceName];
                    var oldValue = oldPrefs[preferenceName];
                    if (newValue === undefined && oldValue !== newValue
                        || oldValue === undefined && newValue !== oldValue // JSONExt.deepEqual() does not support handling `undefined`
                        || !coreutils_1.JSONExt.deepEqual(oldValue, newValue)) {
                        acceptChange({ newValue: newValue, oldValue: oldValue, preferenceName: preferenceName, scope: preference_scope_1.PreferenceScope.Workspace, domain: [] });
                    }
                }
            }
            catch (e_5_1) { e_5 = { error: e_5_1 }; }
            finally {
                try {
                    if (_j && !_j.done && (_c = _h.return)) _c.call(_h);
                }
                finally { if (e_5) throw e_5.error; }
            }
            this.preferences = newPrefs;
        }
        // emit the changes
        var changedPreferenceNames = Object.keys(changesToEmit);
        if (changedPreferenceNames.length > 0) {
            this.onPreferencesChangedEmitter.fire(changesToEmit);
        }
        changedPreferenceNames.forEach(function (preferenceName) { return _this.onPreferenceChangedEmitter.fire(changesToEmit[preferenceName]); });
    };
    PreferenceServiceImpl.prototype.getAffectedPreferenceNames = function (change, accept) {
        var e_6, _a;
        accept(change.preferenceName);
        try {
            for (var _b = __values(this.schema.getOverridePreferenceNames(change.preferenceName)), _c = _b.next(); !_c.done; _c = _b.next()) {
                var overridePreferenceName = _c.value;
                if (!this.doHas(overridePreferenceName)) {
                    accept(overridePreferenceName);
                }
            }
        }
        catch (e_6_1) { e_6 = { error: e_6_1 }; }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
            }
            finally { if (e_6) throw e_6.error; }
        }
    };
    PreferenceServiceImpl.prototype.doCreateProvider = function (scope) {
        if (!this.providersMap.has(scope)) {
            var provider = this.providerProvider(scope);
            this.doSetProvider(scope, provider);
            return provider;
        }
        return this.providersMap.get(scope);
    };
    PreferenceServiceImpl.prototype.doSetProvider = function (scope, provider) {
        this.providersMap.set(scope, provider);
        this.providers.push(provider);
        this.toDispose.push(provider);
    };
    PreferenceServiceImpl.prototype.getProvider = function (scope, preferenceName, resourceUri) {
        var provider = this.providersMap.get(scope);
        if (provider && (!preferenceName || provider.canProvide(preferenceName, resourceUri).priority >= 0)) {
            return provider;
        }
        return undefined;
    };
    PreferenceServiceImpl.prototype.getPreferences = function (resourceUri) {
        var e_7, _a;
        var preferences = {};
        try {
            for (var _b = __values(this.schema.getPreferenceNames()), _c = _b.next(); !_c.done; _c = _b.next()) {
                var preferenceName = _c.value;
                preferences[preferenceName] = this.get(preferenceName, undefined, resourceUri);
            }
        }
        catch (e_7_1) { e_7 = { error: e_7_1 }; }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
            }
            finally { if (e_7) throw e_7.error; }
        }
        return preferences;
    };
    PreferenceServiceImpl.prototype.has = function (preferenceName, resourceUri) {
        return this.get(preferenceName, undefined, resourceUri) !== undefined;
    };
    PreferenceServiceImpl.prototype.get = function (preferenceName, defaultValue, resourceUri) {
        var value = this.doGet(preferenceName, defaultValue, resourceUri);
        if (value === null || value === undefined) {
            var overriden = this.overridenPreferenceName(preferenceName);
            if (overriden) {
                return this.doGet(overriden.preferenceName, defaultValue, resourceUri);
            }
        }
        return value;
    };
    PreferenceServiceImpl.prototype.set = function (preferenceName, value, scope, resourceUri) {
        if (scope === void 0) { scope = preference_scope_1.PreferenceScope.User; }
        return __awaiter(this, void 0, void 0, function () {
            var provider;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        provider = this.getProvider(scope);
                        if (!provider) return [3 /*break*/, 2];
                        return [4 /*yield*/, provider.setPreference(preferenceName, value, resourceUri)];
                    case 1:
                        _a.sent();
                        _a.label = 2;
                    case 2: return [2 /*return*/];
                }
            });
        });
    };
    PreferenceServiceImpl.prototype.getBoolean = function (preferenceName, defaultValue, resourceUri) {
        var value = resourceUri ? this.get(preferenceName, defaultValue, resourceUri) : this.get(preferenceName, defaultValue);
        return value !== null && value !== undefined ? !!value : defaultValue;
    };
    PreferenceServiceImpl.prototype.getString = function (preferenceName, defaultValue, resourceUri) {
        var value = resourceUri ? this.get(preferenceName, defaultValue, resourceUri) : this.get(preferenceName, defaultValue);
        if (value === null || value === undefined) {
            return defaultValue;
        }
        return value.toString();
    };
    PreferenceServiceImpl.prototype.getNumber = function (preferenceName, defaultValue, resourceUri) {
        var value = resourceUri ? this.get(preferenceName, defaultValue, resourceUri) : this.get(preferenceName, defaultValue);
        if (value === null || value === undefined) {
            return defaultValue;
        }
        if (typeof value === 'number') {
            return value;
        }
        return Number(value);
    };
    PreferenceServiceImpl.prototype.inspect = function (preferenceName, resourceUri) {
        var schemaProps = this.schema.getPreferenceProperty(preferenceName);
        if (schemaProps) {
            var defaultValue = schemaProps.default;
            var globalValue = this.inspectInScope(preferenceName, preference_scope_1.PreferenceScope.User, resourceUri);
            var workspaceValue = this.inspectInScope(preferenceName, preference_scope_1.PreferenceScope.Workspace, resourceUri);
            var workspaceFolderValue = this.inspectInScope(preferenceName, preference_scope_1.PreferenceScope.Folder, resourceUri);
            return { preferenceName: preferenceName, defaultValue: defaultValue, globalValue: globalValue, workspaceValue: workspaceValue, workspaceFolderValue: workspaceFolderValue };
        }
    };
    PreferenceServiceImpl.prototype.inspectInScope = function (preferenceName, scope, resourceUri) {
        var value = this.doInspectInScope(preferenceName, scope, resourceUri);
        if (value === null || value === undefined) {
            var overriden = this.overridenPreferenceName(preferenceName);
            if (overriden) {
                return this.doInspectInScope(overriden.preferenceName, scope, resourceUri);
            }
        }
        return value;
    };
    PreferenceServiceImpl.prototype.overridePreferenceName = function (options) {
        return this.schema.overridePreferenceName(options);
    };
    PreferenceServiceImpl.prototype.overridenPreferenceName = function (preferenceName) {
        return this.schema.overridenPreferenceName(preferenceName);
    };
    PreferenceServiceImpl.prototype.doHas = function (preferenceName, resourceUri) {
        return this.doGet(preferenceName, undefined, resourceUri) !== undefined;
    };
    PreferenceServiceImpl.prototype.doInspectInScope = function (preferenceName, scope, resourceUri) {
        var provider = this.getProvider(scope, preferenceName, resourceUri);
        return provider && provider.get(preferenceName, resourceUri);
    };
    PreferenceServiceImpl.prototype.doGet = function (preferenceName, defaultValue, resourceUri) {
        var e_8, _a;
        try {
            for (var _b = __values(preference_scope_1.PreferenceScope.getReversedScopes()), _c = _b.next(); !_c.done; _c = _b.next()) {
                var scope = _c.value;
                if (this.schema.isValidInScope(preferenceName, scope)) {
                    var provider = this.getProvider(scope, preferenceName, resourceUri);
                    if (provider) {
                        var value = provider.get(preferenceName, resourceUri);
                        var result = value !== null && value !== undefined ? value : defaultValue;
                        return common_1.deepFreeze(result);
                    }
                }
            }
        }
        catch (e_8_1) { e_8 = { error: e_8_1 }; }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
            }
            finally { if (e_8) throw e_8.error; }
        }
    };
    __decorate([
        inversify_1.inject(preference_contribution_1.PreferenceSchemaProvider),
        __metadata("design:type", preference_contribution_1.PreferenceSchemaProvider)
    ], PreferenceServiceImpl.prototype, "schema", void 0);
    __decorate([
        inversify_1.inject(exports.PreferenceProviderProvider),
        __metadata("design:type", Function)
    ], PreferenceServiceImpl.prototype, "providerProvider", void 0);
    __decorate([
        inversify_1.postConstruct(),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", []),
        __metadata("design:returntype", void 0)
    ], PreferenceServiceImpl.prototype, "init", null);
    PreferenceServiceImpl = __decorate([
        inversify_1.injectable()
    ], PreferenceServiceImpl);
    return PreferenceServiceImpl;
}());
exports.PreferenceServiceImpl = PreferenceServiceImpl;
