// @theia/plugin-ext/lib/main/browser/plugin-contribution-handler.js
"use strict";
/********************************************************************************
 * Copyright (C) 2018 Red Hat, Inc. and others.
 *
 * This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License v. 2.0 which is available at
 * http://www.eclipse.org/legal/epl-2.0.
 *
 * This Source Code may also be made available under the following Secondary
 * Licenses when the conditions for such availability set forth in the Eclipse
 * Public License v. 2.0 are satisfied: GNU General Public License, version 2
 * with the GNU Classpath Exception which is available at
 * https://www.gnu.org/software/classpath/license.html.
 *
 * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
 ********************************************************************************/
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __values = (this && this.__values) || function (o) {
    var m = typeof Symbol === "function" && o[Symbol.iterator], i = 0;
    if (m) return m.call(o);
    return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
};
Object.defineProperty(exports, "__esModule", { value: true });
var inversify_1 = require("inversify");
var textmate_1 = require("@theia/monaco/lib/browser/textmate");
var menus_contribution_handler_1 = require("./menus/menus-contribution-handler");
var view_registry_1 = require("./view/view-registry");
var browser_1 = require("@theia/core/lib/browser");
var keybindings_contribution_handler_1 = require("./keybindings/keybindings-contribution-handler");
var monaco_snippet_suggest_provider_1 = require("@theia/monaco/lib/browser/monaco-snippet-suggest-provider");
var plugin_shared_style_1 = require("./plugin-shared-style");
var core_1 = require("@theia/core");
var theming_1 = require("@theia/core/lib/browser/theming");
var PluginContributionHandler = /** @class */ (function () {
    function PluginContributionHandler() {
        this.injections = new Map();
        this.pluginCommandIconId = 0;
    }
    PluginContributionHandler.prototype.handleContributions = function (contributions) {
        var _this = this;
        var e_1, _a, e_2, _b, e_3, _c;
        if (contributions.configuration) {
            this.updateConfigurationSchema(contributions.configuration);
        }
        if (contributions.configurationDefaults) {
            this.updateDefaultOverridesSchema(contributions.configurationDefaults);
        }
        if (contributions.languages) {
            try {
                for (var _d = __values(contributions.languages), _e = _d.next(); !_e.done; _e = _d.next()) {
                    var lang = _e.value;
                    monaco.languages.register({
                        id: lang.id,
                        aliases: lang.aliases,
                        extensions: lang.extensions,
                        filenamePatterns: lang.filenamePatterns,
                        filenames: lang.filenames,
                        firstLine: lang.firstLine,
                        mimetypes: lang.mimetypes
                    });
                    if (lang.configuration) {
                        monaco.languages.setLanguageConfiguration(lang.id, {
                            wordPattern: this.createRegex(lang.configuration.wordPattern),
                            autoClosingPairs: lang.configuration.autoClosingPairs,
                            brackets: lang.configuration.brackets,
                            comments: lang.configuration.comments,
                            folding: this.convertFolding(lang.configuration.folding),
                            surroundingPairs: lang.configuration.surroundingPairs,
                            indentationRules: this.convertIndentationRules(lang.configuration.indentationRules)
                        });
                    }
                }
            }
            catch (e_1_1) { e_1 = { error: e_1_1 }; }
            finally {
                try {
                    if (_e && !_e.done && (_a = _d.return)) _a.call(_d);
                }
                finally { if (e_1) throw e_1.error; }
            }
        }
        if (contributions.grammars) {
            var _loop_1 = function (grammar) {
                var e_4, _a;
                if (grammar.injectTo) {
                    try {
                        for (var _b = __values(grammar.injectTo), _c = _b.next(); !_c.done; _c = _b.next()) {
                            var injectScope = _c.value;
                            var injections = this_1.injections.get(injectScope);
                            if (!injections) {
                                injections = [];
                                this_1.injections.set(injectScope, injections);
                            }
                            injections.push(grammar.scope);
                        }
                    }
                    catch (e_4_1) { e_4 = { error: e_4_1 }; }
                    finally {
                        try {
                            if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
                        }
                        finally { if (e_4) throw e_4.error; }
                    }
                }
                this_1.grammarsRegistry.registerTextmateGrammarScope(grammar.scope, {
                    getGrammarDefinition: function () {
                        return __awaiter(this, void 0, void 0, function () {
                            return __generator(this, function (_a) {
                                return [2 /*return*/, {
                                        format: grammar.format,
                                        content: grammar.grammar || '',
                                    }];
                            });
                        });
                    },
                    getInjections: function (scopeName) {
                        return _this.injections.get(scopeName);
                    }
                });
                if (grammar.language) {
                    this_1.grammarsRegistry.mapLanguageIdToTextmateGrammar(grammar.language, grammar.scope);
                    this_1.grammarsRegistry.registerGrammarConfiguration(grammar.language, {
                        embeddedLanguages: this_1.convertEmbeddedLanguages(grammar.embeddedLanguages),
                        tokenTypes: this_1.convertTokenTypes(grammar.tokenTypes)
                    });
                    monaco.languages.onLanguage(grammar.language, function () { return _this.monacoTextmateService.activateLanguage(grammar.language); });
                }
            };
            var this_1 = this;
            try {
                for (var _f = __values(contributions.grammars), _g = _f.next(); !_g.done; _g = _f.next()) {
                    var grammar = _g.value;
                    _loop_1(grammar);
                }
            }
            catch (e_2_1) { e_2 = { error: e_2_1 }; }
            finally {
                try {
                    if (_g && !_g.done && (_b = _f.return)) _b.call(_f);
                }
                finally { if (e_2) throw e_2.error; }
            }
        }
        if (contributions.viewsContainers) {
            var _loop_2 = function (location_1) {
                if (contributions.viewsContainers.hasOwnProperty(location_1)) {
                    var viewContainers = contributions.viewsContainers[location_1];
                    viewContainers.forEach(function (container) {
                        var views = contributions.views && contributions.views[container.id] ? contributions.views[container.id] : [];
                        _this.viewRegistry.registerViewContainer(location_1, container, views);
                    });
                }
            };
            for (var location_1 in contributions.viewsContainers) {
                _loop_2(location_1);
            }
        }
        this.registerCommands(contributions);
        this.menusContributionHandler.handle(contributions);
        this.keybindingsContributionHandler.handle(contributions);
        if (contributions.snippets) {
            try {
                for (var _h = __values(contributions.snippets), _j = _h.next(); !_j.done; _j = _h.next()) {
                    var snippet = _j.value;
                    this.snippetSuggestProvider.fromURI(snippet.uri, {
                        language: snippet.language,
                        source: snippet.source
                    });
                }
            }
            catch (e_3_1) { e_3 = { error: e_3_1 }; }
            finally {
                try {
                    if (_j && !_j.done && (_c = _h.return)) _c.call(_h);
                }
                finally { if (e_3) throw e_3.error; }
            }
        }
    };
    PluginContributionHandler.prototype.registerCommands = function (contribution) {
        var e_5, _a;
        if (!contribution.commands) {
            return;
        }
        var _loop_3 = function (iconUrl, command, category, title) {
            var iconClass = void 0;
            if (iconUrl) {
                iconClass = 'plugin-command-icon-' + this_2.pluginCommandIconId++;
                var darkIconUrl_1 = typeof iconUrl === 'object' ? iconUrl.dark : iconUrl;
                var lightIconUrl_1 = typeof iconUrl === 'object' ? iconUrl.light : iconUrl;
                this_2.style.insertRule('.' + iconClass, function (theme) { return "\n                    width: 16px;\n                    height: 16px;\n                    background: no-repeat url(\"" + (theme.id === theming_1.BuiltinThemeProvider.lightTheme.id ? lightIconUrl_1 : darkIconUrl_1) + "\");\n                "; });
            }
            this_2.commands.registerCommand({
                id: command,
                category: category,
                label: title,
                iconClass: iconClass
            });
        };
        var this_2 = this;
        try {
            for (var _b = __values(contribution.commands), _c = _b.next(); !_c.done; _c = _b.next()) {
                var _d = _c.value, iconUrl = _d.iconUrl, command = _d.command, category = _d.category, title = _d.title;
                _loop_3(iconUrl, command, category, title);
            }
        }
        catch (e_5_1) { e_5 = { error: e_5_1 }; }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
            }
            finally { if (e_5) throw e_5.error; }
        }
    };
    PluginContributionHandler.prototype.updateConfigurationSchema = function (schema) {
        this.validateConfigurationSchema(schema);
        this.preferenceSchemaProvider.setSchema(schema);
    };
    PluginContributionHandler.prototype.updateDefaultOverridesSchema = function (configurationDefaults) {
        var defaultOverrides = {
            id: 'defaultOverrides',
            title: 'Default Configuration Overrides',
            properties: {}
        };
        // tslint:disable-next-line:forin
        for (var key in configurationDefaults) {
            var defaultValue = configurationDefaults[key];
            if (this.preferenceSchemaProvider.testOverrideValue(key, defaultValue)) {
                defaultOverrides.properties[key] = {
                    type: 'object',
                    default: defaultValue,
                    description: "Configure editor settings to be overridden for " + key + " language."
                };
            }
        }
        if (Object.keys(defaultOverrides.properties).length) {
            this.preferenceSchemaProvider.setSchema(defaultOverrides);
        }
    };
    PluginContributionHandler.prototype.createRegex = function (value) {
        if (typeof value === 'string') {
            return new RegExp(value, '');
        }
        return undefined;
    };
    PluginContributionHandler.prototype.convertIndentationRules = function (rules) {
        if (!rules) {
            return undefined;
        }
        return {
            decreaseIndentPattern: this.createRegex(rules.decreaseIndentPattern),
            increaseIndentPattern: this.createRegex(rules.increaseIndentPattern),
            indentNextLinePattern: this.createRegex(rules.indentNextLinePattern),
            unIndentedLinePattern: this.createRegex(rules.unIndentedLinePattern)
        };
    };
    PluginContributionHandler.prototype.convertFolding = function (folding) {
        if (!folding) {
            return undefined;
        }
        var result = {
            offSide: folding.offSide
        };
        if (folding.markers) {
            result.markers = {
                end: this.createRegex(folding.markers.end),
                start: this.createRegex(folding.markers.start)
            };
        }
        return result;
    };
    PluginContributionHandler.prototype.convertTokenTypes = function (tokenTypes) {
        if (typeof tokenTypes === 'undefined' || tokenTypes === null) {
            return undefined;
        }
        // tslint:disable-next-line:no-null-keyword
        var result = Object.create(null);
        var scopes = Object.keys(tokenTypes);
        var len = scopes.length;
        for (var i = 0; i < len; i++) {
            var scope = scopes[i];
            var tokenType = tokenTypes[scope];
            switch (tokenType) {
                case 'string':
                    result[scope] = 2 /* String */;
                    break;
                case 'other':
                    result[scope] = 0 /* Other */;
                    break;
                case 'comment':
                    result[scope] = 1 /* Comment */;
                    break;
            }
        }
        return result;
    };
    PluginContributionHandler.prototype.convertEmbeddedLanguages = function (languages) {
        if (typeof languages === 'undefined' || languages === null) {
            return undefined;
        }
        // tslint:disable-next-line:no-null-keyword
        var result = Object.create(null);
        var scopes = Object.keys(languages);
        var len = scopes.length;
        for (var i = 0; i < len; i++) {
            var scope = scopes[i];
            var langId = languages[scope];
            result[scope] = textmate_1.getEncodedLanguageId(langId);
        }
        return result;
    };
    PluginContributionHandler.prototype.validateConfigurationSchema = function (schema) {
        // tslint:disable-next-line:forin
        for (var p in schema.properties) {
            var property = schema.properties[p];
            if (property.type !== 'object') {
                continue;
            }
            if (!property.default) {
                this.validateDefaultValue(property);
            }
            var properties = property['properties'];
            if (properties) {
                // tslint:disable-next-line:forin
                for (var key in properties) {
                    if (typeof properties[key] !== 'object') {
                        delete properties[key];
                    }
                }
            }
        }
    };
    PluginContributionHandler.prototype.validateDefaultValue = function (property) {
        property.default = {};
        var properties = property['properties'];
        if (properties) {
            // tslint:disable-next-line:forin
            for (var key in properties) {
                if (properties[key].default) {
                    property.default[key] = properties[key].default;
                    delete properties[key].default;
                }
            }
        }
    };
    __decorate([
        inversify_1.inject(textmate_1.TextmateRegistry),
        __metadata("design:type", textmate_1.TextmateRegistry)
    ], PluginContributionHandler.prototype, "grammarsRegistry", void 0);
    __decorate([
        inversify_1.inject(view_registry_1.ViewRegistry),
        __metadata("design:type", view_registry_1.ViewRegistry)
    ], PluginContributionHandler.prototype, "viewRegistry", void 0);
    __decorate([
        inversify_1.inject(menus_contribution_handler_1.MenusContributionPointHandler),
        __metadata("design:type", menus_contribution_handler_1.MenusContributionPointHandler)
    ], PluginContributionHandler.prototype, "menusContributionHandler", void 0);
    __decorate([
        inversify_1.inject(browser_1.PreferenceSchemaProvider),
        __metadata("design:type", browser_1.PreferenceSchemaProvider)
    ], PluginContributionHandler.prototype, "preferenceSchemaProvider", void 0);
    __decorate([
        inversify_1.inject(textmate_1.MonacoTextmateService),
        __metadata("design:type", textmate_1.MonacoTextmateService)
    ], PluginContributionHandler.prototype, "monacoTextmateService", void 0);
    __decorate([
        inversify_1.inject(keybindings_contribution_handler_1.KeybindingsContributionPointHandler),
        __metadata("design:type", keybindings_contribution_handler_1.KeybindingsContributionPointHandler)
    ], PluginContributionHandler.prototype, "keybindingsContributionHandler", void 0);
    __decorate([
        inversify_1.inject(monaco_snippet_suggest_provider_1.MonacoSnippetSuggestProvider),
        __metadata("design:type", monaco_snippet_suggest_provider_1.MonacoSnippetSuggestProvider)
    ], PluginContributionHandler.prototype, "snippetSuggestProvider", void 0);
    __decorate([
        inversify_1.inject(core_1.CommandRegistry),
        __metadata("design:type", core_1.CommandRegistry)
    ], PluginContributionHandler.prototype, "commands", void 0);
    __decorate([
        inversify_1.inject(plugin_shared_style_1.PluginSharedStyle),
        __metadata("design:type", plugin_shared_style_1.PluginSharedStyle)
    ], PluginContributionHandler.prototype, "style", void 0);
    PluginContributionHandler = __decorate([
        inversify_1.injectable()
    ], PluginContributionHandler);
    return PluginContributionHandler;
}());
exports.PluginContributionHandler = PluginContributionHandler;
