// @theia/core/lib/browser/shell/tab-bar-toolbar.js
"use strict";
/********************************************************************************
 * Copyright (C) 2018 TypeFox and others.
 *
 * This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License v. 2.0 which is available at
 * http://www.eclipse.org/legal/epl-2.0.
 *
 * This Source Code may also be made available under the following Secondary
 * Licenses when the conditions for such availability set forth in the Eclipse
 * Public License v. 2.0 are satisfied: GNU General Public License, version 2
 * with the GNU Classpath Exception which is available at
 * https://www.gnu.org/software/classpath/license.html.
 *
 * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
 ********************************************************************************/
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    }
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
var __spread = (this && this.__spread) || function () {
    for (var ar = [], i = 0; i < arguments.length; i++) ar = ar.concat(__read(arguments[i]));
    return ar;
};
var __values = (this && this.__values) || function (o) {
    var m = typeof Symbol === "function" && o[Symbol.iterator], i = 0;
    if (m) return m.call(o);
    return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
};
Object.defineProperty(exports, "__esModule", { value: true });
var React = require("react");
var inversify_1 = require("inversify");
var widgets_1 = require("../widgets");
var label_parser_1 = require("../label-parser");
var contribution_provider_1 = require("../../common/contribution-provider");
var command_1 = require("../../common/command");
var disposable_1 = require("../../common/disposable");
var context_key_service_1 = require("../context-key-service");
var event_1 = require("../../common/event");
var debounce = require("lodash.debounce");
/**
 * Factory for instantiating tab-bar toolbars.
 */
exports.TabBarToolbarFactory = Symbol('TabBarToolbarFactory');
/**
 * Tab-bar toolbar widget representing the active [tab-bar toolbar items](TabBarToolbarItem).
 */
var TabBarToolbar = /** @class */ (function (_super) {
    __extends(TabBarToolbar, _super);
    function TabBarToolbar(commands, labelParser) {
        var _this = _super.call(this) || this;
        _this.commands = commands;
        _this.labelParser = labelParser;
        _this.items = new Map();
        _this.executeCommand = function (e) {
            var item = _this.items.get(e.currentTarget.id);
            if (item) {
                _this.commands.executeCommand(item.command, _this.current);
            }
        };
        _this.addClass(TabBarToolbar.Styles.TAB_BAR_TOOLBAR);
        _this.hide();
        return _this;
    }
    TabBarToolbar.prototype.updateItems = function (items, current) {
        var _this = this;
        this.items = new Map(items.sort(TabBarToolbarItem.PRIORITY_COMPARATOR).reverse().map(function (item) { return [item.id, item]; }));
        this.current = current;
        if (!this.items.size) {
            this.hide();
        }
        this.onRender.push(disposable_1.Disposable.create(function () {
            if (_this.items.size) {
                _this.show();
            }
        }));
        this.update();
    };
    TabBarToolbar.prototype.render = function () {
        var _this = this;
        return React.createElement(React.Fragment, null, __spread(this.items.values()).map(function (item) { return _this.renderItem(item); }));
    };
    TabBarToolbar.prototype.renderItem = function (item) {
        var e_1, _a;
        var innerText = '';
        var classNames = [];
        if (item.text) {
            try {
                for (var _b = __values(this.labelParser.parse(item.text)), _c = _b.next(); !_c.done; _c = _b.next()) {
                    var labelPart = _c.value;
                    if (typeof labelPart !== 'string' && label_parser_1.LabelIcon.is(labelPart)) {
                        var className = "fa fa-" + labelPart.name + (labelPart.animation ? ' fa-' + labelPart.animation : '');
                        classNames.push.apply(classNames, __spread(className.split(' ')));
                    }
                    else {
                        innerText = labelPart;
                    }
                }
            }
            catch (e_1_1) { e_1 = { error: e_1_1 }; }
            finally {
                try {
                    if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
                }
                finally { if (e_1) throw e_1.error; }
            }
        }
        var command = this.commands.getCommand(item.command);
        if (command) {
            var iconClass = command.iconClass;
            if (iconClass) {
                classNames.push(iconClass);
            }
        }
        return React.createElement("div", { key: item.id, className: "" + TabBarToolbar.Styles.TAB_BAR_TOOLBAR_ITEM + (command && this.commandIsEnabled(command.id) ? ' enabled' : '') },
            React.createElement("div", { id: item.id, className: classNames.join(' '), onClick: this.executeCommand, title: item.tooltip }, innerText));
    };
    TabBarToolbar.prototype.commandIsEnabled = function (command) {
        return this.commands.isEnabled(command, this.current);
    };
    return TabBarToolbar;
}(widgets_1.ReactWidget));
exports.TabBarToolbar = TabBarToolbar;
(function (TabBarToolbar) {
    var Styles;
    (function (Styles) {
        Styles.TAB_BAR_TOOLBAR = 'p-TabBar-toolbar';
        Styles.TAB_BAR_TOOLBAR_ITEM = 'item';
    })(Styles = TabBarToolbar.Styles || (TabBarToolbar.Styles = {}));
})(TabBarToolbar = exports.TabBarToolbar || (exports.TabBarToolbar = {}));
exports.TabBarToolbar = TabBarToolbar;
/**
 * Clients should implement this interface if they want to contribute to the tab-bar toolbar.
 */
exports.TabBarToolbarContribution = Symbol('TabBarToolbarContribution');
var TabBarToolbarItem;
(function (TabBarToolbarItem) {
    /**
     * Compares the items by `priority` in ascending. Undefined priorities will be treated as `0`.
     */
    TabBarToolbarItem.PRIORITY_COMPARATOR = function (left, right) {
        // The navigation group is special as it will always be sorted to the top/beginning of a menu.
        if (left.group === 'navigation') {
            return -1;
        }
        if (right.group === 'navigation') {
            return 1;
        }
        if (left.group && right.group) {
            if (left.group < right.group) {
                return -1;
            }
            else if (left.group > right.group) {
                return 1;
            }
            else {
                return 0;
            }
        }
        if (left.group) {
            return -1;
        }
        if (right.group) {
            return 1;
        }
        return (left.priority || 0) - (right.priority || 0);
    };
})(TabBarToolbarItem = exports.TabBarToolbarItem || (exports.TabBarToolbarItem = {}));
/**
 * Main, shared registry for tab-bar toolbar items.
 */
var TabBarToolbarRegistry = /** @class */ (function () {
    function TabBarToolbarRegistry() {
        var _this = this;
        this.items = new Map();
        this.onDidChangeEmitter = new event_1.Emitter();
        this.onDidChange = this.onDidChangeEmitter.event;
        // debounce in order to avoid to fire more than once in the same tick
        this.fireOnDidChange = debounce(function () { return _this.onDidChangeEmitter.fire(undefined); }, 0);
    }
    TabBarToolbarRegistry.prototype.onStart = function () {
        var e_2, _a;
        var contributions = this.contributionProvider.getContributions();
        try {
            for (var contributions_1 = __values(contributions), contributions_1_1 = contributions_1.next(); !contributions_1_1.done; contributions_1_1 = contributions_1.next()) {
                var contribution = contributions_1_1.value;
                contribution.registerToolbarItems(this);
            }
        }
        catch (e_2_1) { e_2 = { error: e_2_1 }; }
        finally {
            try {
                if (contributions_1_1 && !contributions_1_1.done && (_a = contributions_1.return)) _a.call(contributions_1);
            }
            finally { if (e_2) throw e_2.error; }
        }
    };
    /**
     * Registers the given item. Throws an error, if the corresponding command cannot be found or an item has been already registered for the desired command.
     *
     * @param item the item to register.
     */
    TabBarToolbarRegistry.prototype.registerItem = function (item) {
        var _this = this;
        var id = item.id;
        if (this.items.has(id)) {
            throw new Error("A toolbar item is already registered with the '" + id + "' ID.");
        }
        this.items.set(id, item);
        this.fireOnDidChange();
        if (item.onDidChange) {
            item.onDidChange(function () { return _this.fireOnDidChange(); });
        }
    };
    /**
     * Returns an array of tab-bar toolbar items which are visible when the `widget` argument is the current one.
     *
     * By default returns with all items where the command is enabled and `item.isVisible` is `true`.
     */
    TabBarToolbarRegistry.prototype.visibleItems = function (widget) {
        var e_3, _a;
        var result = [];
        try {
            for (var _b = __values(this.items.values()), _c = _b.next(); !_c.done; _c = _b.next()) {
                var item = _c.value;
                if (this.commandRegistry.isVisible(item.command, widget) &&
                    (!item.when || this.contextKeyService.match(item.when, widget.node))) {
                    result.push(item);
                }
            }
        }
        catch (e_3_1) { e_3 = { error: e_3_1 }; }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
            }
            finally { if (e_3) throw e_3.error; }
        }
        return result;
    };
    __decorate([
        inversify_1.inject(command_1.CommandRegistry),
        __metadata("design:type", command_1.CommandRegistry)
    ], TabBarToolbarRegistry.prototype, "commandRegistry", void 0);
    __decorate([
        inversify_1.inject(context_key_service_1.ContextKeyService),
        __metadata("design:type", context_key_service_1.ContextKeyService)
    ], TabBarToolbarRegistry.prototype, "contextKeyService", void 0);
    __decorate([
        inversify_1.inject(contribution_provider_1.ContributionProvider),
        inversify_1.named(exports.TabBarToolbarContribution),
        __metadata("design:type", Object)
    ], TabBarToolbarRegistry.prototype, "contributionProvider", void 0);
    TabBarToolbarRegistry = __decorate([
        inversify_1.injectable()
    ], TabBarToolbarRegistry);
    return TabBarToolbarRegistry;
}());
exports.TabBarToolbarRegistry = TabBarToolbarRegistry;
