// @theia/plugin-ext/lib/hosted/node/hosted-plugin-deployer-handler.js
"use strict";
/********************************************************************************
 * Copyright (C) 2019 RedHat and others.
 *
 * This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License v. 2.0 which is available at
 * http://www.eclipse.org/legal/epl-2.0.
 *
 * This Source Code may also be made available under the following Secondary
 * Licenses when the conditions for such availability set forth in the Eclipse
 * Public License v. 2.0 are satisfied: GNU General Public License, version 2
 * with the GNU Classpath Exception which is available at
 * https://www.gnu.org/software/classpath/license.html.
 *
 * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
 ********************************************************************************/
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __values = (this && this.__values) || function (o) {
    var m = typeof Symbol === "function" && o[Symbol.iterator], i = 0;
    if (m) return m.call(o);
    return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
};
Object.defineProperty(exports, "__esModule", { value: true });
var inversify_1 = require("inversify");
var core_1 = require("@theia/core");
var plugin_reader_1 = require("./plugin-reader");
var HostedPluginDeployerHandler = /** @class */ (function () {
    function HostedPluginDeployerHandler() {
        /**
         * Managed plugin metadata backend entries.
         */
        this.currentBackendPluginsMetadata = [];
        /**
         * Managed plugin metadata frontend entries.
         */
        this.currentFrontendPluginsMetadata = [];
    }
    HostedPluginDeployerHandler.prototype.getDeployedFrontendMetadata = function () {
        return this.currentFrontendPluginsMetadata;
    };
    HostedPluginDeployerHandler.prototype.getDeployedBackendMetadata = function () {
        return this.currentBackendPluginsMetadata;
    };
    HostedPluginDeployerHandler.prototype.deployFrontendPlugins = function (frontendPlugins) {
        return __awaiter(this, void 0, void 0, function () {
            var e_1, _a, _loop_1, this_1, frontendPlugins_1, frontendPlugins_1_1, plugin, e_1_1;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        _loop_1 = function (plugin) {
                            var metadata;
                            return __generator(this, function (_a) {
                                switch (_a.label) {
                                    case 0: return [4 /*yield*/, this_1.reader.getPluginMetadata(plugin.path())];
                                    case 1:
                                        metadata = _a.sent();
                                        if (metadata) {
                                            if (this_1.getDeployedFrontendMetadata().some(function (value) { return value.model.id === metadata.model.id; })) {
                                                return [2 /*return*/, "continue"];
                                            }
                                            this_1.currentFrontendPluginsMetadata.push(metadata);
                                            this_1.logger.info("Deploying frontend plugin \"" + metadata.model.name + "@" + metadata.model.version + "\" from \"" + (metadata.model.entryPoint.frontend || plugin.path()) + "\"");
                                        }
                                        return [2 /*return*/];
                                }
                            });
                        };
                        this_1 = this;
                        _b.label = 1;
                    case 1:
                        _b.trys.push([1, 6, 7, 8]);
                        frontendPlugins_1 = __values(frontendPlugins), frontendPlugins_1_1 = frontendPlugins_1.next();
                        _b.label = 2;
                    case 2:
                        if (!!frontendPlugins_1_1.done) return [3 /*break*/, 5];
                        plugin = frontendPlugins_1_1.value;
                        return [5 /*yield**/, _loop_1(plugin)];
                    case 3:
                        _b.sent();
                        _b.label = 4;
                    case 4:
                        frontendPlugins_1_1 = frontendPlugins_1.next();
                        return [3 /*break*/, 2];
                    case 5: return [3 /*break*/, 8];
                    case 6:
                        e_1_1 = _b.sent();
                        e_1 = { error: e_1_1 };
                        return [3 /*break*/, 8];
                    case 7:
                        try {
                            if (frontendPlugins_1_1 && !frontendPlugins_1_1.done && (_a = frontendPlugins_1.return)) _a.call(frontendPlugins_1);
                        }
                        finally { if (e_1) throw e_1.error; }
                        return [7 /*endfinally*/];
                    case 8: return [2 /*return*/];
                }
            });
        });
    };
    HostedPluginDeployerHandler.prototype.deployBackendPlugins = function (backendPlugins) {
        return __awaiter(this, void 0, void 0, function () {
            var e_2, _a, _loop_2, this_2, backendPlugins_1, backendPlugins_1_1, plugin, e_2_1;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        _loop_2 = function (plugin) {
                            var metadata;
                            return __generator(this, function (_a) {
                                switch (_a.label) {
                                    case 0: return [4 /*yield*/, this_2.reader.getPluginMetadata(plugin.path())];
                                    case 1:
                                        metadata = _a.sent();
                                        if (metadata) {
                                            if (this_2.getDeployedBackendMetadata().some(function (value) { return value.model.id === metadata.model.id; })) {
                                                return [2 /*return*/, "continue"];
                                            }
                                            this_2.currentBackendPluginsMetadata.push(metadata);
                                            this_2.logger.info("Deploying backend plugin \"" + metadata.model.name + "@" + metadata.model.version + "\" from \"" + (metadata.model.entryPoint.backend || plugin.path()) + "\"");
                                        }
                                        return [2 /*return*/];
                                }
                            });
                        };
                        this_2 = this;
                        _b.label = 1;
                    case 1:
                        _b.trys.push([1, 6, 7, 8]);
                        backendPlugins_1 = __values(backendPlugins), backendPlugins_1_1 = backendPlugins_1.next();
                        _b.label = 2;
                    case 2:
                        if (!!backendPlugins_1_1.done) return [3 /*break*/, 5];
                        plugin = backendPlugins_1_1.value;
                        return [5 /*yield**/, _loop_2(plugin)];
                    case 3:
                        _b.sent();
                        _b.label = 4;
                    case 4:
                        backendPlugins_1_1 = backendPlugins_1.next();
                        return [3 /*break*/, 2];
                    case 5: return [3 /*break*/, 8];
                    case 6:
                        e_2_1 = _b.sent();
                        e_2 = { error: e_2_1 };
                        return [3 /*break*/, 8];
                    case 7:
                        try {
                            if (backendPlugins_1_1 && !backendPlugins_1_1.done && (_a = backendPlugins_1.return)) _a.call(backendPlugins_1);
                        }
                        finally { if (e_2) throw e_2.error; }
                        return [7 /*endfinally*/];
                    case 8: return [2 /*return*/];
                }
            });
        });
    };
    __decorate([
        inversify_1.inject(core_1.ILogger),
        __metadata("design:type", Object)
    ], HostedPluginDeployerHandler.prototype, "logger", void 0);
    __decorate([
        inversify_1.inject(plugin_reader_1.HostedPluginReader),
        __metadata("design:type", plugin_reader_1.HostedPluginReader)
    ], HostedPluginDeployerHandler.prototype, "reader", void 0);
    HostedPluginDeployerHandler = __decorate([
        inversify_1.injectable()
    ], HostedPluginDeployerHandler);
    return HostedPluginDeployerHandler;
}());
exports.HostedPluginDeployerHandler = HostedPluginDeployerHandler;
