// @theia/plugin-ext/lib/hosted/node/plugin-host-rpc.js
"use strict";
/********************************************************************************
 * Copyright (C) 2018 Red Hat, Inc. and others.
 *
 * This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License v. 2.0 which is available at
 * http://www.eclipse.org/legal/epl-2.0.
 *
 * This Source Code may also be made available under the following Secondary
 * Licenses when the conditions for such availability set forth in the Eclipse
 * Public License v. 2.0 are satisfied: GNU General Public License, version 2
 * with the GNU Classpath Exception which is available at
 * https://www.gnu.org/software/classpath/license.html.
 *
 * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
 ********************************************************************************/
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __values = (this && this.__values) || function (o) {
    var m = typeof Symbol === "function" && o[Symbol.iterator], i = 0;
    if (m) return m.call(o);
    return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
};
Object.defineProperty(exports, "__esModule", { value: true });
var plugin_manager_1 = require("../../plugin/plugin-manager");
var plugin_api_1 = require("../../api/plugin-api");
var plugin_context_1 = require("../../plugin/plugin-context");
var env_1 = require("../../plugin/env");
var preference_registry_1 = require("../../plugin/preference-registry");
var debug_1 = require("../../plugin/node/debug/debug");
var editors_and_documents_1 = require("../../plugin/editors-and-documents");
var workspace_1 = require("../../plugin/workspace");
var message_registry_1 = require("../../plugin/message-registry");
/**
 * Handle the RPC calls.
 */
var PluginHostRPC = /** @class */ (function () {
    // tslint:disable-next-line:no-any
    function PluginHostRPC(rpc) {
        this.rpc = rpc;
    }
    PluginHostRPC.prototype.initialize = function () {
        var envExt = new env_1.EnvExtImpl(this.rpc);
        var debugExt = new debug_1.DebugExtImpl(this.rpc);
        var editorsAndDocumentsExt = new editors_and_documents_1.EditorsAndDocumentsExtImpl(this.rpc);
        var messageRegistryExt = new message_registry_1.MessageRegistryExt(this.rpc);
        var workspaceExt = new workspace_1.WorkspaceExtImpl(this.rpc, editorsAndDocumentsExt, messageRegistryExt);
        var preferenceRegistryExt = new preference_registry_1.PreferenceRegistryExtImpl(this.rpc, workspaceExt);
        this.pluginManager = this.createPluginManager(envExt, preferenceRegistryExt, this.rpc);
        this.rpc.set(plugin_api_1.MAIN_RPC_CONTEXT.HOSTED_PLUGIN_MANAGER_EXT, this.pluginManager);
        this.rpc.set(plugin_api_1.MAIN_RPC_CONTEXT.EDITORS_AND_DOCUMENTS_EXT, editorsAndDocumentsExt);
        this.rpc.set(plugin_api_1.MAIN_RPC_CONTEXT.WORKSPACE_EXT, workspaceExt);
        this.rpc.set(plugin_api_1.MAIN_RPC_CONTEXT.PREFERENCE_REGISTRY_EXT, preferenceRegistryExt);
        this.apiFactory = plugin_context_1.createAPIFactory(this.rpc, this.pluginManager, envExt, debugExt, preferenceRegistryExt, editorsAndDocumentsExt, workspaceExt, messageRegistryExt);
    };
    // tslint:disable-next-line:no-any
    PluginHostRPC.prototype.initContext = function (contextPath, plugin) {
        console.log('PLUGIN_HOST(' + process.pid + '): initializing(' + contextPath + ')');
        try {
            var backendInit = require(contextPath);
            backendInit.doInitialization(this.apiFactory, plugin);
        }
        catch (e) {
            console.error(e);
        }
    };
    /*
     * Stop the given context by calling the plug-in manager.
     * note: stopPlugin can also be invoked through RPC proxy.
     */
    PluginHostRPC.prototype.stopContext = function () {
        return this.pluginManager.$stopPlugin('');
    };
    // tslint:disable-next-line:no-any
    PluginHostRPC.prototype.createPluginManager = function (envExt, preferencesManager, rpc) {
        var _this = this;
        var extensionTestsPath = process.env.extensionTestsPath;
        var self = this;
        var pluginManager = new plugin_manager_1.PluginManagerExtImpl({
            loadPlugin: function (plugin) {
                console.log('PLUGIN_HOST(' + process.pid + '): PluginManagerExtImpl/loadPlugin(' + plugin.pluginPath + ')');
                try {
                    // cleaning the cache for all files of that plug-in.
                    Object.keys(require.cache).forEach(function (key) {
                        if (key.startsWith(plugin.pluginFolder)) {
                            // delete entry
                            delete require.cache[key];
                        }
                    });
                    return require(plugin.pluginPath);
                }
                catch (e) {
                    console.error(e);
                }
            },
            init: function (raw) {
                var e_1, _a;
                console.log('PLUGIN_HOST(' + process.pid + '): PluginManagerExtImpl/init()');
                var result = [];
                var foreign = [];
                try {
                    for (var raw_1 = __values(raw), raw_1_1 = raw_1.next(); !raw_1_1.done; raw_1_1 = raw_1.next()) {
                        var plg = raw_1_1.value;
                        var pluginModel = plg.model;
                        var pluginLifecycle = plg.lifecycle;
                        if (pluginModel.entryPoint.backend) {
                            var backendInitPath = pluginLifecycle.backendInitPath;
                            // if no init path, try to init as regular Theia plugin
                            if (!backendInitPath) {
                                backendInitPath = __dirname + '/scanners/backend-init-theia.js';
                            }
                            var plugin = {
                                pluginPath: pluginModel.entryPoint.backend,
                                pluginFolder: plg.source.packagePath,
                                model: pluginModel,
                                lifecycle: pluginLifecycle,
                                rawModel: plg.source
                            };
                            self.initContext(backendInitPath, plugin);
                            result.push(plugin);
                        }
                        else {
                            foreign.push({
                                pluginPath: pluginModel.entryPoint.frontend,
                                pluginFolder: plg.source.packagePath,
                                model: pluginModel,
                                lifecycle: pluginLifecycle,
                                rawModel: plg.source
                            });
                        }
                    }
                }
                catch (e_1_1) { e_1 = { error: e_1_1 }; }
                finally {
                    try {
                        if (raw_1_1 && !raw_1_1.done && (_a = raw_1.return)) _a.call(raw_1);
                    }
                    finally { if (e_1) throw e_1.error; }
                }
                return [result, foreign];
            },
            initExtApi: function (extApi) {
                var e_2, _a;
                try {
                    for (var extApi_1 = __values(extApi), extApi_1_1 = extApi_1.next(); !extApi_1_1.done; extApi_1_1 = extApi_1.next()) {
                        var api = extApi_1_1.value;
                        if (api.backendInitPath) {
                            try {
                                var extApiInit = require(api.backendInitPath);
                                extApiInit.provideApi(rpc, pluginManager);
                            }
                            catch (e) {
                                console.error(e);
                            }
                        }
                    }
                }
                catch (e_2_1) { e_2 = { error: e_2_1 }; }
                finally {
                    try {
                        if (extApi_1_1 && !extApi_1_1.done && (_a = extApi_1.return)) _a.call(extApi_1);
                    }
                    finally { if (e_2) throw e_2.error; }
                }
            },
            loadTests: extensionTestsPath ? function () { return __awaiter(_this, void 0, void 0, function () {
                var testRunner, requireError;
                return __generator(this, function (_a) {
                    try {
                        testRunner = require(extensionTestsPath);
                    }
                    catch (error) {
                        requireError = error;
                    }
                    // Execute the runner if it follows our spec
                    if (testRunner && typeof testRunner.run === 'function') {
                        return [2 /*return*/, new Promise(function (resolve, reject) {
                                testRunner.run(extensionTestsPath, function (error) {
                                    if (error) {
                                        reject(error.toString());
                                    }
                                    else {
                                        resolve(undefined);
                                    }
                                });
                            })];
                    }
                    throw new Error(requireError ?
                        requireError.toString() :
                        "Path " + extensionTestsPath + " does not point to a valid extension test runner.");
                });
            }); } : undefined
        }, envExt, preferencesManager, rpc);
        return pluginManager;
    };
    return PluginHostRPC;
}());
exports.PluginHostRPC = PluginHostRPC;
