// @theia/file-search/lib/node/file-search-service-impl.js
"use strict";
/********************************************************************************
 * Copyright (C) 2017 TypeFox and others.
 *
 * This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License v. 2.0 which is available at
 * http://www.eclipse.org/legal/epl-2.0.
 *
 * This Source Code may also be made available under the following Secondary
 * Licenses when the conditions for such availability set forth in the Eclipse
 * Public License v. 2.0 are satisfied: GNU General Public License, version 2
 * with the GNU Classpath Exception which is available at
 * https://www.gnu.org/software/classpath/license.html.
 *
 * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
 ********************************************************************************/
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
var __spread = (this && this.__spread) || function () {
    for (var ar = [], i = 0; i < arguments.length; i++) ar = ar.concat(__read(arguments[i]));
    return ar;
};
var __values = (this && this.__values) || function (o) {
    var m = typeof Symbol === "function" && o[Symbol.iterator], i = 0;
    if (m) return m.call(o);
    return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
};
Object.defineProperty(exports, "__esModule", { value: true });
var fuzzy = require("fuzzy");
var readline = require("readline");
var vscode_ripgrep_1 = require("vscode-ripgrep");
var inversify_1 = require("inversify");
var uri_1 = require("@theia/core/lib/common/uri");
var file_uri_1 = require("@theia/core/lib/node/file-uri");
var core_1 = require("@theia/core");
var node_1 = require("@theia/process/lib/node");
var FileSearchServiceImpl = /** @class */ (function () {
    function FileSearchServiceImpl(logger, rawProcessFactory) {
        this.logger = logger;
        this.rawProcessFactory = rawProcessFactory;
    }
    FileSearchServiceImpl.prototype.find = function (searchPattern, options, clientToken) {
        return __awaiter(this, void 0, void 0, function () {
            var cancellationSource, token, opts, exactMatches, fuzzyMatches, stringPattern;
            var _this = this;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        cancellationSource = new core_1.CancellationTokenSource();
                        if (clientToken) {
                            clientToken.onCancellationRequested(function () { return cancellationSource.cancel(); });
                        }
                        token = cancellationSource.token;
                        if (options.defaultIgnorePatterns && options.defaultIgnorePatterns.length === 0) { // default excludes
                            options.defaultIgnorePatterns.push('.git');
                        }
                        opts = __assign({ fuzzyMatch: true, limit: Number.MAX_SAFE_INTEGER, useGitIgnore: true }, options);
                        exactMatches = new Set();
                        fuzzyMatches = new Set();
                        stringPattern = searchPattern.toLocaleLowerCase();
                        return [4 /*yield*/, Promise.all(options.rootUris.map(function (root) { return __awaiter(_this, void 0, void 0, function () {
                                var rootUri_1, e_1;
                                return __generator(this, function (_a) {
                                    switch (_a.label) {
                                        case 0:
                                            _a.trys.push([0, 2, , 3]);
                                            rootUri_1 = new uri_1.default(root);
                                            return [4 /*yield*/, this.doFind(rootUri_1, searchPattern, opts, function (candidate) {
                                                    var fileUri = rootUri_1.resolve(candidate).toString();
                                                    if (exactMatches.has(fileUri) || fuzzyMatches.has(fileUri)) {
                                                        return;
                                                    }
                                                    if (!searchPattern || searchPattern === '*' || candidate.toLocaleLowerCase().indexOf(stringPattern) !== -1) {
                                                        exactMatches.add(fileUri);
                                                    }
                                                    else if (opts.fuzzyMatch && fuzzy.test(searchPattern, candidate)) {
                                                        fuzzyMatches.add(fileUri);
                                                    }
                                                    if (exactMatches.size + fuzzyMatches.size === opts.limit) {
                                                        cancellationSource.cancel();
                                                    }
                                                }, token)];
                                        case 1:
                                            _a.sent();
                                            return [3 /*break*/, 3];
                                        case 2:
                                            e_1 = _a.sent();
                                            console.error('Failed to search:', root, e_1);
                                            return [3 /*break*/, 3];
                                        case 3: return [2 /*return*/];
                                    }
                                });
                            }); }))];
                    case 1:
                        _a.sent();
                        if (clientToken && clientToken.isCancellationRequested) {
                            return [2 /*return*/, []];
                        }
                        return [2 /*return*/, __spread(exactMatches, fuzzyMatches)];
                }
            });
        });
    };
    FileSearchServiceImpl.prototype.doFind = function (rootUri, searchPattern, options, accept, token) {
        var _this = this;
        return new Promise(function (resolve, reject) {
            try {
                var cwd = file_uri_1.FileUri.fsPath(rootUri);
                var args = _this.getSearchArgs(options);
                // TODO: why not just child_process.spawn, theia process are supposed to be used for user processes like tasks and terminals, not internal
                var process_1 = _this.rawProcessFactory({ command: vscode_ripgrep_1.rgPath, args: args, options: { cwd: cwd } });
                process_1.onError(reject);
                process_1.output.on('close', resolve);
                token.onCancellationRequested(function () { return process_1.kill(); });
                var lineReader = readline.createInterface({
                    input: process_1.output,
                    output: process_1.input
                });
                lineReader.on('line', function (line) {
                    if (token.isCancellationRequested) {
                        process_1.kill();
                    }
                    else {
                        accept(line);
                    }
                });
            }
            catch (e) {
                reject(e);
            }
        });
    };
    FileSearchServiceImpl.prototype.getSearchArgs = function (options) {
        var e_2, _a;
        var args = ['--files', '--case-sensitive'];
        if (options.includePatterns) {
            try {
                for (var _b = __values(options.includePatterns), _c = _b.next(); !_c.done; _c = _b.next()) {
                    var includePattern = _c.value;
                    var includeGlob = includePattern;
                    if (!includeGlob.endsWith('*')) {
                        includeGlob = includeGlob + "*";
                    }
                    if (!includeGlob.startsWith('*')) {
                        includeGlob = "*" + includeGlob;
                    }
                    args.push('--glob', includeGlob);
                }
            }
            catch (e_2_1) { e_2 = { error: e_2_1 }; }
            finally {
                try {
                    if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
                }
                finally { if (e_2) throw e_2.error; }
            }
        }
        if (!options.useGitIgnore) {
            args.push('-uu');
        }
        if (options && options.defaultIgnorePatterns) {
            options.defaultIgnorePatterns.filter(function (p) { return p !== ''; })
                .forEach(function (ignore) {
                if (!ignore.endsWith('*')) {
                    ignore = ignore + "*";
                }
                if (!ignore.startsWith('*')) {
                    ignore = "!*" + ignore;
                }
                else {
                    ignore = "!" + ignore;
                }
                args.push('--glob');
                args.push(ignore);
            });
        }
        return args;
    };
    FileSearchServiceImpl = __decorate([
        inversify_1.injectable(),
        __param(0, inversify_1.inject(core_1.ILogger)),
        __param(1, inversify_1.inject(node_1.RawProcessFactory)),
        __metadata("design:paramtypes", [Object, Function])
    ], FileSearchServiceImpl);
    return FileSearchServiceImpl;
}());
exports.FileSearchServiceImpl = FileSearchServiceImpl;
