// @theia/plugin-ext/lib/main/browser/keybindings/keybindings-contribution-handler.js
"use strict";
/********************************************************************************
 * Copyright (C) 2018 Red Hat, Inc. and others.
 *
 * This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License v. 2.0 which is available at
 * http://www.eclipse.org/legal/epl-2.0.
 *
 * This Source Code may also be made available under the following Secondary
 * Licenses when the conditions for such availability set forth in the Eclipse
 * Public License v. 2.0 are satisfied: GNU General Public License, version 2
 * with the GNU Classpath Exception which is available at
 * https://www.gnu.org/software/classpath/license.html.
 *
 * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
 ********************************************************************************/
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __values = (this && this.__values) || function (o) {
    var m = typeof Symbol === "function" && o[Symbol.iterator], i = 0;
    if (m) return m.call(o);
    return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
};
Object.defineProperty(exports, "__esModule", { value: true });
var inversify_1 = require("inversify");
var keybinding_1 = require("@theia/core/lib/browser/keybinding");
var browser_1 = require("@theia/core/lib/browser");
var logger_1 = require("@theia/core/lib/common/logger");
var os_1 = require("@theia/core/lib/common/os");
var KeybindingsContributionPointHandler = /** @class */ (function () {
    function KeybindingsContributionPointHandler() {
    }
    KeybindingsContributionPointHandler.prototype.handle = function (contributions) {
        var e_1, _a;
        if (!contributions || !contributions.keybindings) {
            return;
        }
        var keybindings = [];
        try {
            for (var _b = __values(contributions.keybindings), _c = _b.next(); !_c.done; _c = _b.next()) {
                var raw = _c.value;
                var keybinding = this.toKeybinding(raw);
                if (keybinding) {
                    try {
                        var keybindingResult = this.keybindingRegistry.getKeybindingsForKeySequence(browser_1.KeySequence.parse(keybinding.keybinding));
                        this.handleShadingKeybindings(keybinding, keybindingResult.shadow);
                        this.handlePartialKeybindings(keybinding, keybindingResult.partial);
                        keybindings.push(keybinding);
                    }
                    catch (e) {
                        this.logger.error(e.message || e);
                    }
                }
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
            }
            finally { if (e_1) throw e_1.error; }
        }
        this.keybindingRegistry.setKeymap(keybinding_1.KeybindingScope.USER, keybindings);
    };
    KeybindingsContributionPointHandler.prototype.toKeybinding = function (pluginKeybinding) {
        var keybinding = this.toOSKeybinding(pluginKeybinding);
        if (!keybinding) {
            return undefined;
        }
        var command = pluginKeybinding.command, when = pluginKeybinding.when;
        return { keybinding: keybinding, command: command, when: when };
    };
    KeybindingsContributionPointHandler.prototype.toOSKeybinding = function (pluginKeybinding) {
        var keybinding;
        var os = os_1.OS.type();
        if (os === os_1.OS.Type.Windows) {
            keybinding = pluginKeybinding.win;
        }
        else if (os === os_1.OS.Type.OSX) {
            keybinding = pluginKeybinding.mac;
        }
        else {
            keybinding = pluginKeybinding.linux;
        }
        return keybinding || pluginKeybinding.keybinding;
    };
    KeybindingsContributionPointHandler.prototype.handlePartialKeybindings = function (keybinding, partialKeybindings) {
        var _this = this;
        partialKeybindings.forEach(function (partial) {
            if (keybinding.context === undefined || keybinding.context === partial.context) {
                _this.logger.warn("Partial keybinding is ignored; " + keybinding_1.Keybinding.stringify(keybinding) + " shadows " + keybinding_1.Keybinding.stringify(partial));
            }
        });
    };
    KeybindingsContributionPointHandler.prototype.handleShadingKeybindings = function (keybinding, shadingKeybindings) {
        var _this = this;
        shadingKeybindings.forEach(function (shadow) {
            if (shadow.context === undefined || shadow.context === keybinding.context) {
                _this.keybindingRegistry.unregisterKeybinding(shadow);
                _this.logger.warn("Shadowing keybinding is ignored; " + keybinding_1.Keybinding.stringify(shadow) + ", shadows " + keybinding_1.Keybinding.stringify(keybinding));
            }
        });
    };
    __decorate([
        inversify_1.inject(logger_1.ILogger),
        __metadata("design:type", Object)
    ], KeybindingsContributionPointHandler.prototype, "logger", void 0);
    __decorate([
        inversify_1.inject(keybinding_1.KeybindingRegistry),
        __metadata("design:type", keybinding_1.KeybindingRegistry)
    ], KeybindingsContributionPointHandler.prototype, "keybindingRegistry", void 0);
    KeybindingsContributionPointHandler = __decorate([
        inversify_1.injectable()
    ], KeybindingsContributionPointHandler);
    return KeybindingsContributionPointHandler;
}());
exports.KeybindingsContributionPointHandler = KeybindingsContributionPointHandler;
