// @theia/task/lib/browser/task-configurations.js
"use strict";
/********************************************************************************
 * Copyright (C) 2017-2018 Ericsson and others.
 *
 * This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License v. 2.0 which is available at
 * http://www.eclipse.org/legal/epl-2.0.
 *
 * This Source Code may also be made available under the following Secondary
 * Licenses when the conditions for such availability set forth in the Eclipse
 * Public License v. 2.0 are satisfied: GNU General Public License, version 2
 * with the GNU Classpath Exception which is available at
 * https://www.gnu.org/software/classpath/license.html.
 *
 * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
 ********************************************************************************/
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __values = (this && this.__values) || function (o) {
    var m = typeof Symbol === "function" && o[Symbol.iterator], i = 0;
    if (m) return m.call(o);
    return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
};
Object.defineProperty(exports, "__esModule", { value: true });
var inversify_1 = require("inversify");
var common_1 = require("@theia/core/lib/common");
var uri_1 = require("@theia/core/lib/common/uri");
var filesystem_watcher_1 = require("@theia/filesystem/lib/browser/filesystem-watcher");
var filesystem_watcher_protocol_1 = require("@theia/filesystem/lib/common/filesystem-watcher-protocol");
var common_2 = require("@theia/filesystem/lib/common");
var jsoncparser = require("jsonc-parser");
/**
 * Watches a tasks.json configuration file and provides a parsed version of the contained task configurations
 */
var TaskConfigurations = /** @class */ (function () {
    function TaskConfigurations(watcherServer, fileSystem) {
        var _this = this;
        this.watcherServer = watcherServer;
        this.fileSystem = fileSystem;
        this.toDispose = new common_1.DisposableCollection();
        /**
         * Map of source (path of root folder that the task config comes from) and task config map.
         * For the inner map (i.e., task config map), the key is task label and value TaskConfiguration
         */
        this.tasksMap = new Map();
        this.watchedConfigFileUris = [];
        this.watchersMap = new Map(); // map of watchers for task config files, where the key is folder uri
        /** last directory element under which we look for task config */
        this.TASKFILEPATH = '.theia';
        /** task configuration file name */
        this.TASKFILE = 'tasks.json';
        this.client = undefined;
        this.toDispose.push(watcherServer);
        this.watcherServer.onFilesChanged(function (changes) { return __awaiter(_this, void 0, void 0, function () {
            var watchedConfigFileChanges, err_1;
            var _this = this;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 3, , 4]);
                        watchedConfigFileChanges = changes.filter(function (change) {
                            return _this.watchedConfigFileUris.some(function (fileUri) { return filesystem_watcher_1.FileChangeEvent.isAffected([change], new uri_1.default(fileUri)); });
                        }).map(function (relatedChange) { return ({ uri: relatedChange.uri.toString(), type: relatedChange.type }); });
                        if (!(watchedConfigFileChanges.length >= 0)) return [3 /*break*/, 2];
                        return [4 /*yield*/, this.onDidTaskFileChange(watchedConfigFileChanges)];
                    case 1:
                        _a.sent();
                        if (this.client) {
                            this.client.taskConfigurationChanged(this.getTaskLabels());
                        }
                        _a.label = 2;
                    case 2: return [3 /*break*/, 4];
                    case 3:
                        err_1 = _a.sent();
                        console.error(err_1);
                        return [3 /*break*/, 4];
                    case 4: return [2 /*return*/];
                }
            });
        }); });
        this.toDispose.push(common_1.Disposable.create(function () {
            _this.tasksMap.clear();
            _this.client = undefined;
        }));
    }
    TaskConfigurations.prototype.setClient = function (client) {
        this.client = client;
    };
    TaskConfigurations.prototype.dispose = function () {
        this.toDispose.dispose();
    };
    Object.defineProperty(TaskConfigurations.prototype, "configFileUris", {
        get: function () {
            return this.watchedConfigFileUris;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * Triggers the watching of a potential task configuration file, under the given root URI.
     * Returns whether a configuration file was found.
     */
    TaskConfigurations.prototype.watchConfigurationFile = function (rootUri) {
        return __awaiter(this, void 0, void 0, function () {
            var configFileUri, disposableWatcher_1, disposable;
            var _this = this;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        configFileUri = this.getConfigFileUri(rootUri);
                        if (!!this.watchedConfigFileUris.some(function (uri) { return uri === configFileUri; })) return [3 /*break*/, 2];
                        this.watchedConfigFileUris.push(configFileUri);
                        return [4 /*yield*/, this.watcherServer.watchFileChanges(new uri_1.default(configFileUri))];
                    case 1:
                        disposableWatcher_1 = _a.sent();
                        disposable = common_1.Disposable.create(function () {
                            disposableWatcher_1.dispose();
                            _this.watchersMap.delete(configFileUri);
                            var ind = _this.watchedConfigFileUris.findIndex(function (uri) { return uri === configFileUri; });
                            if (ind >= 0) {
                                _this.watchedConfigFileUris.splice(ind, 1);
                            }
                        });
                        this.watchersMap.set(configFileUri, disposable);
                        this.toDispose.push(disposable);
                        this.refreshTasks(configFileUri);
                        _a.label = 2;
                    case 2: return [4 /*yield*/, this.fileSystem.exists(configFileUri)];
                    case 3:
                        if (_a.sent()) {
                            return [2 /*return*/, true];
                        }
                        else {
                            console.info("Config file " + this.TASKFILE + " does not exist under " + rootUri);
                            return [2 /*return*/, false];
                        }
                        return [2 /*return*/];
                }
            });
        });
    };
    /**
     * Stops watchers added to a potential task configuration file.
     * Returns whether a configuration file was being watched before this function gets called.
     */
    TaskConfigurations.prototype.unwatchConfigurationFile = function (configFileUri) {
        if (!this.watchersMap.has(configFileUri)) {
            return false;
        }
        this.watchersMap.get(configFileUri).dispose();
        return true;
    };
    /** returns the list of known task labels */
    TaskConfigurations.prototype.getTaskLabels = function () {
        return Array.from(this.tasksMap.values()).reduce(function (acc, labelConfigMap) { return acc.concat(Array.from(labelConfigMap.keys())); }, []);
    };
    /** returns the list of known tasks */
    TaskConfigurations.prototype.getTasks = function () {
        return Array.from(this.tasksMap.values()).reduce(function (acc, labelConfigMap) { return acc.concat(Array.from(labelConfigMap.values())); }, []);
    };
    /** returns the task configuration for a given label or undefined if none */
    TaskConfigurations.prototype.getTask = function (source, taskLabel) {
        var labelConfigMap = this.tasksMap.get(source);
        if (labelConfigMap) {
            return labelConfigMap.get(taskLabel);
        }
    };
    /** removes tasks configured in the given task config file */
    TaskConfigurations.prototype.removeTasks = function (configFileUri) {
        var source = this.getSourceFolderFromConfigUri(configFileUri);
        this.tasksMap.delete(source);
    };
    /** returns the string uri of where the config file would be, if it existed under a given root directory */
    TaskConfigurations.prototype.getConfigFileUri = function (rootDir) {
        return new uri_1.default(rootDir).resolve(this.TASKFILEPATH).resolve(this.TASKFILE).toString();
    };
    /**
     * Called when a change, to a config file we watch, is detected.
     * Triggers a reparse, if appropriate.
     */
    TaskConfigurations.prototype.onDidTaskFileChange = function (fileChanges) {
        return __awaiter(this, void 0, void 0, function () {
            var e_1, _a, fileChanges_1, fileChanges_1_1, change, e_1_1;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        _b.trys.push([0, 6, 7, 8]);
                        fileChanges_1 = __values(fileChanges), fileChanges_1_1 = fileChanges_1.next();
                        _b.label = 1;
                    case 1:
                        if (!!fileChanges_1_1.done) return [3 /*break*/, 5];
                        change = fileChanges_1_1.value;
                        if (!(change.type === filesystem_watcher_protocol_1.FileChangeType.DELETED)) return [3 /*break*/, 2];
                        this.removeTasks(change.uri);
                        return [3 /*break*/, 4];
                    case 2: 
                    // re-parse the config file
                    return [4 /*yield*/, this.refreshTasks(change.uri)];
                    case 3:
                        // re-parse the config file
                        _b.sent();
                        _b.label = 4;
                    case 4:
                        fileChanges_1_1 = fileChanges_1.next();
                        return [3 /*break*/, 1];
                    case 5: return [3 /*break*/, 8];
                    case 6:
                        e_1_1 = _b.sent();
                        e_1 = { error: e_1_1 };
                        return [3 /*break*/, 8];
                    case 7:
                        try {
                            if (fileChanges_1_1 && !fileChanges_1_1.done && (_a = fileChanges_1.return)) _a.call(fileChanges_1);
                        }
                        finally { if (e_1) throw e_1.error; }
                        return [7 /*endfinally*/];
                    case 8: return [2 /*return*/];
                }
            });
        });
    };
    /**
     * Tries to read the tasks from a config file and if it successes then updates the list of available tasks.
     * If reading a config file wasn't successful then does nothing.
     */
    TaskConfigurations.prototype.refreshTasks = function (configFileUri) {
        return __awaiter(this, void 0, void 0, function () {
            var e_2, _a, tasksConfigsArray, newTaskMap, tasksConfigsArray_1, tasksConfigsArray_1_1, task, source;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0: return [4 /*yield*/, this.readTasks(configFileUri)];
                    case 1:
                        tasksConfigsArray = _b.sent();
                        if (tasksConfigsArray) {
                            // only clear tasks map when successful at parsing the config file
                            // this way we avoid clearing and re-filling it multiple times if the
                            // user is editing the file in the auto-save mode, having momentarily
                            // non-parsing JSON.
                            this.removeTasks(configFileUri);
                            if (tasksConfigsArray.length > 0) {
                                newTaskMap = new Map();
                                try {
                                    for (tasksConfigsArray_1 = __values(tasksConfigsArray), tasksConfigsArray_1_1 = tasksConfigsArray_1.next(); !tasksConfigsArray_1_1.done; tasksConfigsArray_1_1 = tasksConfigsArray_1.next()) {
                                        task = tasksConfigsArray_1_1.value;
                                        newTaskMap.set(task.label, task);
                                    }
                                }
                                catch (e_2_1) { e_2 = { error: e_2_1 }; }
                                finally {
                                    try {
                                        if (tasksConfigsArray_1_1 && !tasksConfigsArray_1_1.done && (_a = tasksConfigsArray_1.return)) _a.call(tasksConfigsArray_1);
                                    }
                                    finally { if (e_2) throw e_2.error; }
                                }
                                source = this.getSourceFolderFromConfigUri(configFileUri);
                                this.tasksMap.set(source, newTaskMap);
                            }
                        }
                        return [2 /*return*/];
                }
            });
        });
    };
    /** parses a config file and extracts the tasks launch configurations */
    TaskConfigurations.prototype.readTasks = function (uri) {
        return __awaiter(this, void 0, void 0, function () {
            var e_3, _a, response, strippedContent, errors, tasks, errors_1, errors_1_1, e, err_2;
            var _this = this;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0: return [4 /*yield*/, this.fileSystem.exists(uri)];
                    case 1:
                        if (!!(_b.sent())) return [3 /*break*/, 2];
                        return [2 /*return*/, undefined];
                    case 2:
                        _b.trys.push([2, 4, , 5]);
                        return [4 /*yield*/, this.fileSystem.resolveContent(uri)];
                    case 3:
                        response = _b.sent();
                        strippedContent = jsoncparser.stripComments(response.content);
                        errors = [];
                        tasks = jsoncparser.parse(strippedContent, errors);
                        if (errors.length) {
                            try {
                                for (errors_1 = __values(errors), errors_1_1 = errors_1.next(); !errors_1_1.done; errors_1_1 = errors_1.next()) {
                                    e = errors_1_1.value;
                                    console.error("Error parsing " + uri + ": error: " + e.error + ", length:  " + e.length + ", offset:  " + e.offset);
                                }
                            }
                            catch (e_3_1) { e_3 = { error: e_3_1 }; }
                            finally {
                                try {
                                    if (errors_1_1 && !errors_1_1.done && (_a = errors_1.return)) _a.call(errors_1);
                                }
                                finally { if (e_3) throw e_3.error; }
                            }
                        }
                        else {
                            return [2 /*return*/, this.filterDuplicates(tasks['tasks']).map(function (t) { return Object.assign(t, { _source: _this.getSourceFolderFromConfigUri(uri) }); })];
                        }
                        return [3 /*break*/, 5];
                    case 4:
                        err_2 = _b.sent();
                        console.error("Error(s) reading config file: " + uri);
                        return [3 /*break*/, 5];
                    case 5: return [2 /*return*/];
                }
            });
        });
    };
    TaskConfigurations.prototype.filterDuplicates = function (tasks) {
        var e_4, _a;
        var filteredTasks = [];
        var _loop_1 = function (task) {
            if (filteredTasks.some(function (t) { return t.label === task.label; })) {
                // TODO: create a problem marker so that this issue will be visible in the editor?
                console.error("Error parsing " + this_1.TASKFILE + ": found duplicate entry for label: " + task.label);
            }
            else {
                filteredTasks.push(task);
            }
        };
        var this_1 = this;
        try {
            for (var tasks_1 = __values(tasks), tasks_1_1 = tasks_1.next(); !tasks_1_1.done; tasks_1_1 = tasks_1.next()) {
                var task = tasks_1_1.value;
                _loop_1(task);
            }
        }
        catch (e_4_1) { e_4 = { error: e_4_1 }; }
        finally {
            try {
                if (tasks_1_1 && !tasks_1_1.done && (_a = tasks_1.return)) _a.call(tasks_1);
            }
            finally { if (e_4) throw e_4.error; }
        }
        return filteredTasks;
    };
    TaskConfigurations.prototype.getSourceFolderFromConfigUri = function (configFileUri) {
        return new uri_1.default(configFileUri).parent.parent.path.toString();
    };
    TaskConfigurations = __decorate([
        inversify_1.injectable(),
        __param(0, inversify_1.inject(filesystem_watcher_1.FileSystemWatcher)),
        __param(1, inversify_1.inject(common_2.FileSystem)),
        __metadata("design:paramtypes", [filesystem_watcher_1.FileSystemWatcher, Object])
    ], TaskConfigurations);
    return TaskConfigurations;
}());
exports.TaskConfigurations = TaskConfigurations;
